### PR TITLE
Add support for authentication with mercurial repositories.

### PR DESCRIPTION
**Pull request by Elendev**
This PR aim to resolve the issue #7463.

I've added a `Composer\Util\Hg` class similar to the `Composer\Util\Git` class. The purpose of this class is to add the credentials (username and password) at the beginning of the mercurial's URL to allow authentication with the `auth.json` file.
An URL like `https://my-domain.com/my-repo` will be transformed to `https://my-username:my-password@my-domain.com/my-repo` if authentication is required.
The `Composer\Util\Hg` class works in the same way as the `Composer\Util\Git`, providing a `runCommand` method that handle the potential URL modification.

### CODE REVIEW

File: src/Composer/Util/Hg.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        $command = call_user_func($commandCallable, $url);
+
+        if (0 === $this->process->execute($command, $ignoredOutput, $cwd)){
+            return;
+        }
+
+        // Try with the authentication informations available
+        if (preg_match('{^(https?)://((.+)(?:\:(.+))?@)?([^/]+)(/.*)?}mi', $url, $match) && $this->io->hasAuthentication($match[5])) {
+            $auth = $this->io->getAuthentication($match[5]);
+            $authenticatedUrl = $match[1] . '://' . rawurldecode($auth['username']) . ':' . rawurlencode($auth['password']) . '@' . $match[5] . (!empty($match[6])? $match[6]: null);
```
Comment (Seldaek) [inline src/Composer/Util/Hg.php]: It seems odd that you're doing rawurl**decode** on the username. Should that not also be an encode?

File: src/Composer/Util/Hg.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        }, $message);
+    }
+
+    private function throwException($message, $url)
+    {
+        if (0 !== $this->process->execute('hg --version', $ignoredOutput)) {
+            throw new \RuntimeException(self::sanitizeUrl('Failed to clone ' . $url . ', hg was not found, check that it is installed and in your PATH env.' . "\n\n" . $this->process->getErrorOutput()));
+        }
+
+        throw new \RuntimeException(self::sanitizeUrl($message));
```
Comment (stof) [inline src/Composer/Util/Hg.php]: this is not including the process output anymore, which means you don't see the hg error anymore (which can be very useful to understand why it failed and fix it)

File: src/Composer/Util/Hg.php (hunk 1 above)
Comment (Elendev) [reply src/Composer/Util/Hg.php]: You are totally right, nice catch, thank you ! An autocomplete mistake :+1:

File: src/Composer/Util/Hg.php (hunk 2 above)
Comment (Elendev) [reply src/Composer/Util/Hg.php]: You are right. The git version is not so verbose, but as soon as the feature was here before, I'll keep it here

**Comment by Seldaek** [comment]
Thanks

**Pull request merged**